src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs

```diff
@@ -0,0 +1,463 @@
+                throw new ArgumentNullException(nameof(subjectName));
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+            if (string.IsNullOrEmpty(hashAlgorithm.Name))
+                throw new ArgumentException(SR.Cryptography_HashAlgorithmNameNullOrEmpty, nameof(hashAlgorithm));
+
+            SubjectName = new X500DistinguishedName(subjectName);
+
+            _key = key;
+            _generator = X509SignatureGenerator.CreateForECDsa(key);
+            PublicKey = _generator.PublicKey;
+            HashAlgorithm = hashAlgorithm;
+        }
+
+        /// <summary>
+        /// Create a CertificateRequest for the specified subject name, ECDSA key, and hash algorithm.
+        /// </summary>
+        /// <param name="subjectName">
+        ///   The parsed representation of the subject name for the certificate or certificate request.
+        /// </param>
+        /// <param name="key">
+        ///   An ECDSA key whose public key material will be included in the certificate or certificate request.
+        ///   This key will be used as a private key if <see cref="CreateSelfSigned" /> is called.
+        /// </param>
+        /// <param name="hashAlgorithm">
+        ///   The hash algorithm to use when signing the certificate or certificate request.
+        /// </param>
+        public CertificateRequest(X500DistinguishedName subjectName, ECDsa key, HashAlgorithmName hashAlgorithm)
+        {
+            if (subjectName == null)
+                throw new ArgumentNullException(nameof(subjectName));
+            if (key == null)
+                throw new ArgumentNullException(nameof(key));
+            if (string.IsNullOrEmpty(hashAlgorithm.Name))
+                throw new ArgumentException(SR.Cryptography_HashAlgorithmNameNullOrEmpty, nameof(hashAlgorithm));
+
+            SubjectName = subjectName;
+
+            _key = key;
+            _generator = X509SignatureGenerator.CreateForECDsa(key);
+            PublicKey = _generator.PublicKey;
+            HashAlgorithm = hashAlgorithm;
+        }
+
+        /// <summary>
+        /// Create a CertificateRequest for the specified subject name, RSA key, and hash algorithm.
+        /// </summary>
+        /// <param name="subjectName">
+        ///   The string representation of the subject name for the certificate or certificate request.
+        /// </param>
+        /// <param name="key">
+        ///   An RSA key whose public key material will be included in the certificate or certificate request.
+        ///   This key will be used as a private key if <see cref="CreateSelfSigned" /> is called.
+        /// </param>
+        /// <param name="hashAlgorithm">
+        ///   The hash algorithm to use when signing the certificate or certificate request.
+        /// </param>
+        /// <seealso cref="X500DistinguishedName(string)"/>
+        public CertificateRequest(string subjectName, RSA key, HashAlgorithmName hashAlgorithm)
```

Comment:
The problem is the ArgumentNullException for subjectName; since a ctor-deferral has to happen before parameter validation.  Expression-throw could solve it, but I don't know that we've really started using that anywhere.